Fix issue with empty arrays, selecting by id (#14255)
crate/crate: docs/appendices/release-notes/unreleased.rst

- Fixed a regression introduced in 5.3.0 which caused ``INSERT INTO`` statements
  with a ``ON CONFLICT`` clause on tables with generated primary key columns to
  fail with an ``ArrayIndexOutOfBoundsException``.

- Added a workaround for an issue that caused an exception to be thrown when
  inserting a non-array value into a column that is dynamically created by
  inserting an empty array, ultimately modifying the type of the column and then

>   selecting this column by the row's id, for example::

Reviewer: How about
```suggestion
  selecting this column by the row's primary key, for example::
```

    CREATE TABLE t (id int primary key, o OBJECT(dynamic));
    INSERT INTO t VALUES (1, {x=[]});
    INSERT INTO t VALUES (2, {x={}});  /* this is the culprit statement, inserting an object onto an array typed column */

>     SELECT * FROM test WHERE id=1;
>     SQLParseException[Cannot cast object element `thisfieldshouldbeanarray` with value `[]` to type `object`]

Reviewer: ```suggestion
    SELECT * FROM t WHERE id=1;
    SQLParseException[Cannot cast object element `x` with value `[]` to type `object`]
```

  after the fix::

    SELECT * FROM t WHERE id=1;
    +----+-------------+
    | id | o           |
    +----+-------------+

>     |  1 | {"x": null} |

Reviewer: ```suggestion
    |  1 | {"x": null} |
    |  2 | {"x": {}}   |
```

Author: that's not valid, we ask by id, `id = 1` so only one row is returned.

    UNION
    SELECT i, avg(k), j FROM t3 GROUP BY 1, 3
    LIMIT 10

> - Added a workaround for an issue that caused an exception to be thrown when

Reviewer 2: Why is this a workaround from the user perspective? If so, we would have to explain why we think this is a workaround, I think.
```suggestion
- Fixed an issue that caused an exception to be thrown when
```

    SELECT i, avg(k), j FROM t3 GROUP BY 1, 3
    LIMIT 10

> <<<<<<< HEAD

Reviewer 2: rebase traces ;)
